File: packages/sdk-python/veltrix_sdk/resources/log_entries.py
# -*- coding: utf-8 -*-

from .base import BaseResource


class LogEntriesResource(BaseResource):
    """Handles platform log entries (``/api/logs``)."""
    RESOURCE_PATH = "logs"

    def list(self, page=None, limit=None, level=None, source=None,
             from_date=None, to_date=None, **kwargs):
        """Lists log entries with pagination/filtering (GET /api/logs)."""
        params = {
            "page": page,
            "limit": limit,
            "level": level,
            "source": source,
            "fromDate": from_date,
            "toDate": to_date,
        }
        params = {k: v for k, v in params.items() if v is not None}
        return self._list(params=params, **kwargs)

    def create(self, level, source, message, details=None, **kwargs):
        """Creates a log entry (POST /api/logs)."""
        data = {"level": level, "source": source, "message": message, "details": details}
        data = {k: v for k, v in data.items() if v is not None}
        return self._create(data=data, **kwargs)

    def get(self, log_id, **kwargs):
        """Retrieves a specific log entry by its ID (GET /api/logs/{id})."""
        return self._get(resource_id=log_id, **kwargs)

    def delete(self, log_id, **kwargs):
        """Deletes a log entry by its ID (DELETE /api/logs/{id})."""
        return self._delete(resource_id=log_id, **kwargs)

    def get_sources(self, **kwargs):
        """Retrieves the distinct set of log sources (GET /api/logs/sources)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/sources", **kwargs)

    def get_levels(self, **kwargs):
        """Retrieves the distinct set of log levels (GET /api/logs/levels)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/levels", **kwargs)

File: packages/sdk-js/src/resources/log_entries.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';
import { APIError } from '../errors';

// Define interfaces for Log Entry data structures
type LogLevel = 'info' | 'warn' | 'error' | 'debug';

interface LogEntry {
  id: string; // uuid
  timestamp: string; // ISO Date string
  level: LogLevel;
  source: string;
  message: string;
  details?: string | null;
  customerId: string; // uuid
  createdAt: string; // ISO Date string
  updatedAt: string; // ISO Date string
}

interface CreateLogEntryPayload {
  level: LogLevel;
  source: string;
  message: string;
  details?: string | null;
}

interface ListLogEntriesParams {
    page?: number;
    limit?: number;
    level?: LogLevel;
    source?: string;
    fromDate?: string; // ISO Date string
    toDate?: string; // ISO Date string
}

export class LogEntriesResource extends BaseResource {
  protected readonly RESOURCE_PATH = "logs";

  async list(params?: ListLogEntriesParams, config?: AxiosRequestConfig): Promise<LogEntry[]> {
    // Corresponds to GET /api/logs
    const cleanedParams = Object.entries(params || {})
        .filter(([_, value]) => value !== undefined && value !== null)
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._list<LogEntry[]>(cleanedParams, config);
  }

  async create(payload: CreateLogEntryPayload, config?: AxiosRequestConfig): Promise<LogEntry> {
    // Corresponds to POST /api/logs
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._create<LogEntry>(cleanedPayload, undefined, config);
  }

  async get(logId: string, config?: AxiosRequestConfig): Promise<LogEntry> {
    // Corresponds to GET /api/logs/{id}
    return this._get<LogEntry>(logId, undefined, config);
  }

  async delete(logId: string, config?: AxiosRequestConfig): Promise<{ message: string }> {
    // Corresponds to DELETE /api/logs/{id}
    // API Spec indicates 200 OK with message for delete
     const result = await this._delete<{ message: string } | null>(logId, undefined, config);
     if (result === null) {
        throw new APIError(`DELETE ${this._getPath(logId)} returned null/204, expected 200 with message.`);
    }
    return result;
  }

  async getSources(config?: AxiosRequestConfig): Promise<string[]> {
    // Corresponds to GET /api/logs/sources
    const path = `${this.RESOURCE_PATH}/sources`;
    return this.httpClient.get<string[]>(path, config);
  }

  async getLevels(config?: AxiosRequestConfig): Promise<LogLevel[]> {
    // Corresponds to GET /api/logs/levels
    const path = `${this.RESOURCE_PATH}/levels`;
    return this.httpClient.get<LogLevel[]>(path, config);
  }
}
